projects/core/src/lib/colors/color-palette-editor/color-palette-editor.component.ts

```diff
@@ -0,0 +1,144 @@
+import {CdkVirtualScrollViewport} from '@angular/cdk/scrolling';
+import {ChangeDetectionStrategy, Component, OnInit, Input, Output, EventEmitter, ChangeDetectorRef, ViewChild} from '@angular/core';
+import {RasterSymbology} from '../../layers/symbology/symbology.model';
+import {Color} from '../color';
+import {ColorAttributeInput} from '../color-attribute-input/color-attribute-input.component';
+import {ColorBreakpoint} from '../color-breakpoint.model';
+import {PaletteColorizer} from '../colorizer.model';
+
+@Component({
+    selector: 'geoengine-color-palette-editor',
+    templateUrl: './color-palette-editor.component.html',
+    styleUrls: ['./color-palette-editor.component.scss'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class ColorPaletteEditorComponent implements OnInit {
+    // Symbology to use for creating color tabs
+    @Input() symbology!: RasterSymbology;
+    // Symbology altered through color tab inputs
+    @Output() symbologyChanged: EventEmitter<RasterSymbology> = new EventEmitter();
+
+    @ViewChild(CdkVirtualScrollViewport)
+    virtualScrollViewport!: CdkVirtualScrollViewport;
+
+    allColors: ColorAttributeInput[] = new Array<ColorAttributeInput>(); // HTML Template will use these to display the color cards
+    colorMap = new Map<number, Color>(); // Returned to the parent component as parameter for the colorizer
+
+    constructor(private ref: ChangeDetectorRef) {}
+
+    ngOnInit(): void {
+        this.initColorInputs();
+    }
+
+    /**
+     * (re)creates the ColorMap as well as the ColorAttributeInputs based on the current layer
+     */
+    initColorInputs(): void {
+        this.colorMap = new Map<number, Color>();
+        this.allColors = new Array<ColorAttributeInput>();
+
+        this.getPaletteColorizer()
+            .getBreakpoints()
+            .forEach((bp, index) => {
+                const newInput = {key: bp.value.toString(), value: bp.color};
+                this.allColors.push(newInput);
+                this.colorMap.set(bp.value, this.getPaletteColorizer().getColorAtIndex(index));
+            });
+    }
+
+    applyChanges(newSymbology: RasterSymbology): void {
```

Comment:
When you click apply on NDVI, e.g, when you change the opacity, there is a lag before something happens.

Do you know why this lags so badly?

Seems to have something to do with the number of table entries, since it is no problem for land cover.

Maybe it is necessary to add a loading indicator, if this cannot be fixed. Meanwhile, the buttons must be deactivated.

However, I guess there is a chance that this can be fixed.

It could have something to do with the sorting (and redrawing) in `sortColorAttributeInputs`, but I guess you have to debug the slow method first.

Reply:
On the NDVI dataset with 256 colors it takes about 1 second for me, same as with Land Cover. When I added a lot of color tabs, I started to notice higher lag as well. The lag happens after the sorting is done, during redraw of the components (nevertheless I found and took out some unnecessary sorting that was going on).
There's also a big difference between having scrolled the bottom of the list (that is, when all tabs are loaded in) and prior to doing any scrolling.

Long story short, in [3d7cb2d](https://github.com/geo-engine/geoengine-ui/pull/614/commits/3d7cb2d38ad76421a6d50b6679f1232cc5fc3eb1) I removed the [append only mode](https://material.angular.io/cdk/scrolling/overview#append-only-mode) from the virtual viewport, so components that have been scrolled out of view get removed from the DOM. For me this improved things by a lot when having many color tabs, while scrolling up and down quickly is still possible. Hopefully that fixes the issue